### ISSUE TITLE
typeforce/index.js error

### ISSUE THREAD

**Issue by megacrypto** [joshuayabut/z-nomp#15]
i keep getting this error:

> /home/mylin/ZNOMP/z-nomp/node_modules/typeforce/index.js:154
>           throw tfSubError(e, i)
>           ^
> 
> Error: Expected property "1" of type Satoshi, got Number 1174406164.9999998
> 2016-11-29 13:43:24 [Master]    [PoolSpawner] Fork 0 died, spawning replacement worker...
> 2016-11-29 13:43:24 [Switching] [Setup] (Thread 4) Loading last proxy state from redis
> 2016-11-29 13:43:24 [Pool]      [zcash] (Thread 4) Share processing setup with redis (127.0.0.1:6379)
> 2016-11-29 13:43:24 [Switching] [Setup] (Thread 3) Loading last proxy state from redis
> 2016-11-29 13:43:24 [Switching] [Setup] (Thread 4) Switching "switch1" listening for equihash on port 3333 into zcash
> 2016-11-29 13:43:24 [Pool]      [zcash] (Thread 4) No rewardRecipients have been setup which means no fees will be taken
> 2016-11-29 13:43:24 [Pool]      [zcash] (Thread 3) Share processing setup with redis (127.0.0.1:6379)
> 2016-11-29 13:43:24 [Switching] [Setup] (Thread 3) Switching "switch1" listening for equihash on port 3333 into zcash
> 

and it keeps resetting the pool?

**Comment by megacrypto** [comment]
when i turned of the "switching" section in config.json, the error stopped !

**Comment by megacrypto** [comment]
it seems that yes it was because of having the switching (switch 1) enabled!!

**Comment by megacrypto** [comment]
actually error re-appeared again 

> /home/mylin/ZNOMP/z-nomp/node_modules/typeforce/index.js:154
>           throw tfSubError(e, i)
>           ^
> Error: Expected property "1" of type Satoshi, got Number 1177232500.0000002
> 2016-11-29 15:25:00 [Master]    [PoolSpawner] Fork 2 died, spawning replacement worker...
> 2016-11-29 15:25:01 [Switching] [Setup] (Thread 1) Loading last proxy state from redis
> 2016-11-29 15:25:01 [Pool]      [zcash] (Thread 1) Share processing setup with redis (127.0.0.1:6379)
> 2016-11-29 15:25:01 [Pool]      [zcash] (Thread 1) No rewardRecipients have been setup which means no fees will be taken
> 
> /home/mylin/ZNOMP/z-nomp/node_modules/typeforce/index.js:154
>           throw tfSubError(e, i)
>           ^
> Error: Expected property "1" of type Satoshi, got Number 1177232500.0000002
> 2016-11-29 15:25:01 [Master]    [PoolSpawner] Fork 0 died, spawning replacement worker...
> 2016-11-29 15:25:02 [Switching] [Setup] (Thread 2) Loading last proxy state from redis
> 2016-11-29 15:25:02 [Pool]      [zcash] (Thread 2) Share processing setup with redis (127.0.0.1:6379)
> 2016-11-29 15:25:02 [Pool]      [zcash] (Thread 2) No rewardRecipients have been setup which means no fees will be taken
> 
> /home/mylin/ZNOMP/z-nomp/node_modules/typeforce/index.js:154
>           throw tfSubError(e, i)
>           ^
> Error: Expected property "1" of type Satoshi, got Number 1177232500.0000002
> 2016-11-29 15:25:02 [Master]    [PoolSpawner] Fork 1 died, spawning replacement worker...
> 2016-11-29 15:25:02 [Switching] [Setup] (Thread 4) Loading last proxy state from redis
> 2016-11-29 15:25:02 [Pool]      [zcash] (Thread 4) Share processing setup with redis (127.0.0.1:6379)
> 2016-11-29 15:25:02 [Pool]      [zcash] (Thread 4) No rewardRecipients have been setup which means no fees will be taken
> 
> /home/mylin/ZNOMP/z-nomp/node_modules/typeforce/index.js:154
>           throw tfSubError(e, i)
>           ^
> Error: Expected property "1" of type Satoshi, got Number 1177232500.0000002
> 2016-11-29 15:25:02 [Master]    [PoolSpawner] Fork 3 died, spawning replacement worker...
> 2016-11-29 15:25:03 [Switching] [Setup] (Thread 3) Loading last proxy state from redis
> 2016-11-29 15:25:03 [Pool]      [zcash] (Thread 3) Share processing setup with redis (127.0.0.1:6379)
> 2016-11-29 15:25:03 [Pool]      [zcash] (Thread 3) No rewardRecipients have been setup which means no fees will be taken
> 

so, i guess it wasn't the switching part!

**Comment by joshuayabut** [comment]
Can you tell me which version of node that you are running? The error stems from the decimal numbers. 

Anyway as a fix I'll have the number round to the nearest integer.

**Comment by bobfeldbauer** [comment]
I can't tell exactly what's triggering that error, but I still see it now even with your two newest commits to libs/paymentProcessor.js

I'm using nodejs v4.6.1

There's no additional context for the error or stacktrace, but here's what I get in console:

/opt/z-nomp/node_modules/stratum-pool/node_modules/bitcoinjs-lib/node_modules/typeforce/index.js:154
          throw tfSubError(e, i)                                                                            
          ^  

I did try swapping out bitcoinjs-lib-zcash for your bitcoinjs-lib repo, with no change.

**Comment by megacrypto** [comment]
I'm using nodejs v.7.2.0 (as per the instructions) and npm 3.10.9 

I' updating right now and will see if error still exists

**Comment by bobfeldbauer** [comment]
@megacrypto I believe the README is actually referring to node v0.7 and that part hasn't been updated in years since zone117x wrote that in his NOMP repository as far as I know. @joshuayabut  did update the README to use v0.12, but also mentions fixing compatibility with v0.10 nodejs in some of his commits.

Maybe a ~year ago, nodejs changed from 0.x.x versions to integers, going from v0.12.x to v4.0. Basic NOMP has run best in the past on nodejs v0.10.29 due to incompatibility with v0.12 and bignum libs.

**Comment by aayanl** [comment]
Node 12.7 is being referred to as v7 now I think

**Comment by bobfeldbauer** [comment]
@aayanl I don't think it's that direct, but there are still versions numbered v0.12.xx being released. (v7.2.0 and v0.12.17 are both out now, confusingly, and with vastly different node ABI version compatibility...)

**Comment by aayanl** [comment]
im using v7.2 now @bobfeldbauer

**Comment by bobfeldbauer** [comment]
@aayanl Sorry, I meant this issue - not the async error @megacrypto described a couple posts ago.

Typeforce errors like..

`/opt/z-nomp/node_modules/stratum-pool/node_modules/bitcoinjs-lib/node_modules/typeforce/index.js:154
          throw tfSubError(e, i)                                                                            
          ^  
Error: Expected property "1" of type Satoshi, got Number 1202887329.9999998`

**Comment by megacrypto** [comment]
im using ubuntu 16.04 server so i just used the package manager to install it:

https://nodejs.org/en/download/package-manager/#debian-and-ubuntu-based-linux-distributions

**Comment by bobfeldbauer** [comment]
To confirm, I tested with nodejs v7.2.0 and am still seeing this.

```
/opt/z-nomp/node_modules/stratum-pool/node_modules/typeforce/index.js:154
          throw tfSubError(e, i)
          ^
Error: Expected property "1" of type Satoshi, got Number 1235374999.9999998
```

I still haven't found where the initial error is coming from yet, but I think it's probably somewhere in a reward or coinbase calculation, since the value shown is the satoshi amount of the miners+founders rewards. My guess would be from one of the input or output sections in node_modules/stratum-pool/lib/transactions.js

**Comment by joshuayabut** [comment]
I'll have a look at this later tonight or tomorrow. It's an issue where there should be an integer but a float is being returned.

**Comment by megacrypto** [comment]
error does not exist with last git update

**Comment by bobfeldbauer** [comment]
@megacrypto I think you mean the async error was fixed by that commit, not this issue #15 which is about the typeforce error. This issue still exists in current master and maybe shouldn't be closed until it is fixed in the node-stratum-pool repo. I did create a pull request with a fix for it there. https://github.com/joshuayabut/node-stratum-pool/pull/3
